### PR TITLE
Converted ZC_ACK_CHANGE_GUILD_POSITIONINFO to struct

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: 

* **Description of Pull Request**: Fixed sending all positions changed

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -8873,25 +8873,28 @@ void clif_guild_positioninfolist(map_session_data *sd)
 ///     &0x10 = allow expel
 /// ranking:
```
Comment (AoShinRO) [inline src/map/clif.cpp:8874]: What should this ranking be?

File: src/map/packets.hpp [hunk 2]
```diff
@@ -767,6 +767,21 @@ struct PACKET_ZC_NOTIFY_POSITION_TO_GUILDM {
 } __attribute__((packed));
 DEFINE_PACKET_HEADER(ZC_NOTIFY_POSITION_TO_GUILDM, 0x1eb)
 
+struct PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO_sub {
```
Comment (AoShinRO) [inline src/map/packets.hpp:770]: decltype here is copy from ZC_POSITION_INFO packet

File: src/map/clif.cpp [hunk 3]
```diff
@@ -8873,25 +8873,28 @@ void clif_guild_positioninfolist(map_session_data *sd)
...
-	WBUFL(buf,12)=idx;
-	WBUFL(buf,16)=g.position[idx].exp_mode;
-	safestrncpy(WBUFCP(buf,20),g.position[idx].name,NAME_LENGTH);
-	// }*
+	PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO* p = reinterpret_cast<PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO*>( packet_buffer );
+
+	p->PacketType = HEADER_ZC_ACK_CHANGE_GUILD_POSITIONINFO;
+	p->PacketLength = sizeof(*p);
+	int count = 0;
+	for(const guild_position& position : g.position){
```
Comment (AoShinRO) [inline src/map/clif.cpp:8885]:
```suggestion
	for(const guild_position& pos : g.position){
```

Comment (AoShinRO) [inline src/map/clif.cpp:8885]:
```suggestion
	for(const guild_position& pos : g.position){
```

File: src/map/clif.cpp [hunk 4]
```diff
@@ -8873,25 +8873,28 @@ void clif_guild_positioninfolist(map_session_data *sd)
...
+		PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO_sub& Info = p->posInfo[count];
+		Info.positionID = count;
+		Info.mode = position.mode;
```
Comment (AoShinRO) [inline src/map/clif.cpp:8888]:
```suggestion
		Info.mode = pos.mode;
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -8873,25 +8873,28 @@ void clif_guild_positioninfolist(map_session_data *sd)
...
+		Info.ranking = count;
+		Info.payRate = position.exp_mode;
```
Comment (AoShinRO) [inline src/map/clif.cpp:8890]:
```suggestion
		Info.payRate = pos.exp_mode;
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -8873,25 +8873,28 @@ void clif_guild_positioninfolist(map_session_data *sd)
...
+		safestrncpy(Info.posName,position.name,sizeof(Info.posName));
```
Comment (AoShinRO) [inline src/map/clif.cpp:8891]:
```suggestion
		safestrncpy(Info.posName,pos.name,sizeof(Info.posName));
```

File: src/map/guild.cpp [hunk 7]
```diff
@@ -1523,7 +1523,7 @@ int guild_position_changed(int guild_id,int idx,struct guild_position *p) {
 	if(g==nullptr)
 		return 0;
 	memcpy(&g->guild.position[idx],p,sizeof(struct guild_position));
-	clif_guild_positionchanged(g->guild,idx);
+	clif_guild_positionchanged(g->guild);
```
Comment (AoShinRO) [inline src/map/guild.cpp:1526]: maybe we need to change this call location

File: src/map/clif.cpp [hunk 8]
```diff
@@ -8893,25 +8893,28 @@ static void clif_guild_positioninfolist(map_session_data& sd){
...
 ///     TODO
-void clif_guild_positionchanged(const struct mmo_guild &g,int idx)
-{
-	// FIXME: This packet is intended to update the clients after a
-	// commit of position info changes, not sending one packet per
-	// position.
-	map_session_data *sd;
-	unsigned char buf[128];
+void clif_guild_positionchanged(const struct mmo_guild &g){
+	map_session_data *sd = guild_getavailablesd(g);
```
Comment (aleos89) [inline src/map/clif.cpp:8897]:
```suggestion
	map_session_data *sd = guild_getavailablesd(g);

```

File: src/map/clif.cpp [hunk 9]
```diff
@@ -8893,25 +8893,28 @@ static void clif_guild_positioninfolist(map_session_data& sd){
...
-	WBUFL(buf,16)=g.position[idx].exp_mode;
-	safestrncpy(WBUFCP(buf,20),g.position[idx].name,NAME_LENGTH);
-	// }*
-	if( (sd=guild_getavailablesd(g))!=nullptr )
-		clif_send(buf,WBUFW(buf,2),&sd->bl,GUILD);
+	PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO* p = reinterpret_cast<PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO*>( packet_buffer );
+
+	p->PacketType = HEADER_ZC_ACK_CHANGE_GUILD_POSITIONINFO;
+	p->PacketLength = sizeof(*p);
+	int c = 0;
```
Comment (aleos89) [inline src/map/clif.cpp:8905]:
```suggestion

	int c = 0;

```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -8893,25 +8893,28 @@ static void clif_guild_positioninfolist(map_session_data& sd){
...
+	for(const guild_position& gp : g.position){
+		PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO_sub& Info = p->posInfo[c];
+		Info.positionID = c;
+		Info.mode = gp.mode;
+		Info.ranking = c;
+		Info.payRate = gp.exp_mode;
+		safestrncpy(Info.posName,gp.name,sizeof(Info.posName));
+		p->PacketLength += static_cast<decltype(p->PacketLength)>(sizeof(Info));
```
Comment (aleos89) [inline src/map/clif.cpp:8913]:
```suggestion
		PACKET_ZC_ACK_CHANGE_GUILD_POSITIONINFO_sub& info = p->posInfo[c];

		info.positionID = c;
		info.mode = gp.mode;
		info.ranking = c;
		info.payRate = gp.exp_mode;
		safestrncpy(info.posName, gp.name, sizeof(info.posName));
		p->PacketLength += static_cast<decltype(p->PacketLength)>(sizeof(info));
```

**Review by aleos89 — APPROVED** [review]

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request closed without merging**